core/objects.go

```diff
@@ -540,3 +540,33 @@ type SuggestedWindow struct {
 type RenewalInfo struct {
 	SuggestedWindow SuggestedWindow `json:"suggestedWindow"`
 }
+
+// RenewalInfoSimple constructs a `RenewalInfo` object and suggested window
+// using a very simple renewal calculation: Calculate a point 2/3rds of the way
+// through the validity period, then give a 2-day window around that.
+func RenewalInfoSimple(issued time.Time, expires time.Time) RenewalInfo {
+	validity := expires.Add(time.Second).Sub(issued)
+	renewalOffset := time.Duration(int64(0.33 * validity.Seconds()))
+	idealRenewal := expires.UTC().Add(-renewalOffset)
+	return RenewalInfo{
+		SuggestedWindow: SuggestedWindow{
+			Start: idealRenewal.Add(-24 * time.Hour),
+			End:   idealRenewal.Add(24 * time.Hour),
+		},
+	}
+}
+
+// RenewalInfoImmediate constructs a `RenewalInfo` object with a suggested
+// window in the past. The passed `now` is assumed to be a timestamp
+// representing the current moment in time. Per the draft-aaron-ari spec,
+// clients should attempt to renew immediately if the suggested window is in the
+// past.
+func RenewalInfoImmediate(now time.Time) RenewalInfo {
+	oneHourAgo := now.Add(-1 * time.Hour)
+	return RenewalInfo{
+		SuggestedWindow: SuggestedWindow{
+			Start: oneHourAgo,
+			End:   oneHourAgo.Add(time.Minute * 30),
```

Comment:
If the goal is just to say "renew immediately", isn't that adequately expressed by having both the start and end date one hour ago? Does ARI prohibit intervals where the start and end date are the same?

Reply:
The specifications merely states:

```
   suggestedWindow (object, required): A JSON object with two keys,
   "start" and "end", whose values are timestamps, encoded in the format
   specified in [RFC3339], which bound the window of time in which the
   CA recommends renewing the certificate.
```

Reply:
The thought just struck me that the reason I'm not a big fan of a zero duration window is because it's an unexpected result, especially for client authors. For instance, if I write a fairly naive function for deriving a random time between two timestamps a zero duration window will actually result in a panic:

```go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func pickRandomTime(begin, end time.Time) string {
	dur64 := end.Sub(begin).Nanoseconds()
	rand.Seed(time.Now().UnixNano())
	ranDur64 := rand.Int63n(dur64)
	ranDur := time.Duration(ranDur64)
	return begin.Add(ranDur).String()
}

func main() {
	begin := time.Date(2018, 1, 1, 0, 0, 0, 0, time.UTC)
	end := time.Date(2018, 1, 2, 0, 0, 0, 0, time.UTC)
	fmt.Println("non-zero window:\n", pickRandomTime(begin, end))

	end = time.Date(2018, 1, 1, 0, 0, 0, 0, time.UTC)
	fmt.Println("zero window:")
	fmt.Println(pickRandomTime(begin, end))
}

```

```
non-zero window:
 2018-01-01 04:01:40.195147788 +0000 UTC
zero window:
panic: invalid argument to Int63n

goroutine 1 [running]:
math/rand.(*Rand).Int63n(0x0?, 0xed1db7100?)
	/usr/local/go-faketime/src/math/rand/rand.go:110 +0xe8
math/rand.Int63n(...)
	/usr/local/go-faketime/src/math/rand/rand.go:327
main.pickRandomTime({0x4c5098?, 0xc000012018?, 0x0?}, {0x1?, 0x1?, 0x0?})
	/tmp/sandbox1666039483/prog.go:12 +0xde
main.main()
	/tmp/sandbox1666039483/prog.go:24 +0x1b9

Program exited.
```

https://go.dev/play/p/BBHHxefpNSF

Reply:
> I'm not a big fan of a zero duration window is because it's an unexpected result, especially for client authors. For instance, if I write a fairly naive function for deriving a random time between two timestamps a zero duration window will actually result in a panic

Right, this is exactly why I think we should exercise the possibility that the duration window is zero. Otherwise we wind up with a bunch of clients that assume it is never zero (or never think about the possibility that it is zero), which introduces a hidden requirement that's not in the spec.

Going a step further: this code path will be exercised rarely (in incidents). Since we want to make sure clients occasionally see zero duration windows, we should ensure they occasionally see such windows in normal use (renewals). So, for instance, if the recommended renewal window is in the past we could collapse it to a zero duration window instead of giving the +/- 1 day we currently do.

Reply:
Let's discuss this approach offline tomorrow.

Reply:
We've discussed this offline and we're going to land this without zero duration windows and see about categorizing zero duration windows as malformed in the draft specification.